src/fundus/scraping/common_crawl/pipeline.py

```diff
@@ -0,0 +1,274 @@
+
+from fundus.publishers.base_objects import PublisherEnum
+from fundus.scraping.article import Article
+from fundus.scraping.common_crawl.html import CCNewsSource
+from fundus.scraping.common_crawl.scraper import CCNewsScraper
+from fundus.scraping.filter import ExtractionFilter, Requires, URLFilter
+
+_T = TypeVar("_T")
+P = ParamSpec("P")
+
+
+# noinspection PyPep8Naming
+class dill_wrapper(Callable[P, _T]):  # type: ignore[misc]
+    def __init__(self, target: Callable[P, _T]):
+        """Wraps function in dill serialization.
+
+        This is in order to use unpickable functions within multiprocessing.
+
+        Args:
+            target (Callable[P, _T): The function to wrap.
+        """
+        self._serialized_target: bytes = dill.dumps(target)
+
+    @lru_cache
+    def _deserialize(self) -> Callable[P, _T]:
+        return cast(Callable[P, _T], dill.loads(self._serialized_target))
+
+    def __call__(self, *args: P.args, **kwargs: P.kwargs) -> _T:
+        return self._deserialize()(*args, **kwargs)
+
+
+class CCNewsCrawler:
+    def __init__(
+        self,
+        *publishers: PublisherEnum,
+        processes: Optional[int] = None,
+        server_address: str = "https://data.commoncrawl.org/",
+    ):
+        """Initializes a crawler for crawling the CC-NEWS dataset.
+
+        Args:
+            *publishers (PublisherEnum): The publishers to crawl.
+            processes: Number of process to use for crawling. If None, use os.cpu_count(); if -1 omit multiprocessing
+                entirely. Defaults to None
+            server_address: The CC-NEWS dataset server address. Defaults to 'https://data.commoncrawl.org/'
+        """
+        self.publishers = publishers
+        self.processes = processes or os.cpu_count() or -1
+        self.server_address = server_address
+
+    def _get_list_of_warc_path(self, start: datetime, end: datetime) -> List[str]:
+        date_pattern: Pattern[str] = re.compile(r"CC-NEWS-(?P<date>\d{14})-\d{5}")
+
+        if start >= end:
+            raise ValueError("Start date has to be < end date.")
+
+        if start < datetime(2016, 8, 1):
+            raise ValueError("The default, and earliest possible, start date is 2016/08/01.")
+
+        if end > datetime.now():
+            raise ValueError("The specified end date is in the future. We don't want to give spoilers, do we?")
+
+        date_sequence: List[datetime] = [dt for dt in rrule(MONTHLY, dtstart=start, until=end)]
+        urls = [
+            self.server_address + f"crawl-data/CC-NEWS/{date.strftime('%Y/%m')}/warc.paths.gz" for date in date_sequence
+        ]
+
+        def load_paths(url: str) -> List[str]:
+            with requests.Session() as session:
+                paths = gzip.decompress(session.get(url).content).decode("utf-8").split()
+                bar.update()
+                return paths
+
+        with ThreadPool(processes=len(urls)) as pool, tqdm(total=len(urls), desc="Load WARC paths", leave=False) as bar:
```

Comment:
Hmm, it sure is a good idea to limit the number of threads. I would say that two threads per core sounds reasonable.

Reply:
In general this definitely better. But since a good number of threads is different for each system, wouldn't an additional parameter be a good idea? To indicate that this thread parameter would only be used for fetching the WARC paths I think something like `warc_threads` would be good.